org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/nogrammarconstraints/NoGrammarConstraintsCodeAction.java

```diff
@@ -246,18 +249,32 @@ public static TextDocumentEdit createXSINoNamespaceSchemaLocationEdit(String sch
 			throws BadLocationException {
 		String delimiter = document.getTextDocument().lineDelimiter(0);
 		DOMElement documentElement = document.getDocumentElement();
-		int beforeTagOffset = documentElement.getStartTagOpenOffset();
-		int afterTagOffset = beforeTagOffset + 1 + documentElement.getTagName().length();
+		int beforeTagOffset = documentElement != null ? documentElement.getStartTagOpenOffset() : document.getLastChild() != null ? document.getLastChild().getEnd() : 0;
+		int afterTagOffset = documentElement != null ? beforeTagOffset + 1 + documentElement.getTagName().length() : beforeTagOffset;
 
 		StringBuilder insertText = new StringBuilder();
 
+		if (documentElement == null) {
+			if (document.getLastChild() != null) {
+				insertText.append(delimiter);
+			}
+			insertText.append("<");
+			insertText.append(PLACEHOLDER_ELEMENT_NAME);
+		}
+
 		insertText.append(" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"");
 		insertText.append(delimiter);
 
 		insertText.append(" xsi:noNamespaceSchemaLocation=\"");
 		insertText.append(schemaFileName);
 		insertText.append("\"");
 
+		if (documentElement == null) {

```

Comment:
Move this code in a private static method like `generateEndTag(DOMElement documentElement) `to reuse it with other create*Edit

Reply:
I think you should use XMLBuilder every time instead of StringBuilder, You will have the capability to have one method `generateEndTag(DOMElement documentElement, XMLBuilder builder)`